### PR TITLE
Paymaster SC

### PR DESCRIPTION
**Pull request by CostinCarabas**

### CODE REVIEW

File: contracts/paymaster/src/forward_call.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        &self,
+        dest: ManagedAddress,
+        endpoint_name: ManagedBuffer,
+        endpoint_args: MultiValueEncoded<ManagedBuffer>,
+        payments: PaymentsVec<Self::Api>,
+    ) {
+        let original_caller = self.blockchain().get_caller();
+
+        if !self.blockchain().is_smart_contract(&dest) {
+            self.send().direct_multi(&dest, &payments);
```
Comment (sasurobert) [inline contracts/paymaster/src/forward_call.rs:19]: endpoint and arguments - add the message as well in the direct send.

File: contracts/paymaster/src/paymaster.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    fn forward_execution(
+        &self,
+        relayer_addr: ManagedAddress,
+        dest: ManagedAddress,
+        endpoint_name: ManagedBuffer,
+        endpoint_args: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        let payments = self.call_value().all_esdt_transfers();
+
+        self.pay_fee_to_relayer(relayer_addr, payments.clone_value());
```
Comment (sasurobert) [inline contracts/paymaster/src/paymaster.rs:27]: payments.first.clone_calue ? no need to give all the list in this function.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: contracts/paymaster/src/forward_call.rs (hunk 1 above)
Comment (CostinCarabas) [reply contracts/paymaster/src/forward_call.rs]: Created a contract_call for that.
`direct_multi` doesn't support messages anymore.

File: contracts/paymaster/src/paymaster.rs (hunk 2 above)
Comment (CostinCarabas) [reply contracts/paymaster/src/paymaster.rs]: Done

File: contracts/paymaster/src/forward_call.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        &self,
+        dest: ManagedAddress,
+        endpoint_name: ManagedBuffer,
+        endpoint_args: MultiValueEncoded<ManagedBuffer>,
+        payments: PaymentsVec<Self::Api>,
+    ) {
+        let original_caller = self.blockchain().get_caller();
+
+        let (contract_call_endpoint, contract_call_args) =
+            if !self.blockchain().is_smart_contract(&dest) {
```
Comment (sasurobert) [inline contracts/paymaster/src/forward_call.rs:19]: No need for this if. You can do simple transfer to users, but add all arguments received here as data. Or just delete the if statement and do asyncCall as that will resolve it.

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

File: contracts/paymaster/src/paymaster.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        let payments = self.call_value().all_esdt_transfers();
+        require!(!payments.is_empty(), "There is no fee for payment!");
+
+        let fee_payment = payments.get(FEE_PAYMENT);
+        self.send().direct_esdt(
+            &relayer_addr,
+            &fee_payment.token_identifier,
+            0,
+            &fee_payment.amount,
+        );
```
Comment (dorin-iancu) [inline contracts/paymaster/src/paymaster.rs:32]: Shouldn't you check that the fee is equal to some amount?

Comment (sasurobert) [reply contracts/paymaster/src/paymaster.rs:32]: that is checked at the relayer level.

**Review by dorin-iancu — APPROVED** [review]

**Pull request merged**